alf/algorithms/muzero_algorithm.py

```diff
@@ -295,70 +295,182 @@ def _hook(grad, name):
+                replay_buffer.get_current_position().max() - 1)
+
+            # [B, 1]
+            folded_env_ids = start_env_ids.unsqueeze(-1)
+
+            # [B, 1, 1]
+            env_ids = folded_env_ids.unsqueeze(-1)
+            # [B, T, R + 1]
+            positions = folded_positions.unfold(1, R + 1, 1)
+
+            # [B, T]
+            steps_to_episode_end = replay_buffer.steps_to_episode_end(
+                positions[:, :, 0], env_ids[:, :, 0])
+
+            # [B, T]
+            episode_end_positions = positions[:, :, 0] + steps_to_episode_end
+            # [B, T, 1]
+            episode_end_positions = episode_end_positions.unsqueeze(-1)
+
+            # [B, T, R + 1]
+            beyond_episode_end = positions > episode_end_positions
+            positions = torch.min(positions, episode_end_positions)
+
+            # 1. Some times we would like to make sure that a sequence (e.g.
+            #    reanalyze unrolled sequence) to stop at the episode end, and
+            #    elements in the sequence beyond the episode end is filled with
+            #    the value of the the episode end.
+            #
+            # 2. After the unfolding, the original sequence along dim=1 is
+            #    unfolded into T sequences of length R. If the original sequence
+            #    cross the boundaries of the episodes, the new unfolded
+            #    sequences may start at different episodes, and may therefore
+            #    have different episode ends.
+            #
+            # With the above in mind, the helper function below achieves
+            # unfolding and adapting on the episode ends individually for each
+            # unfolded sequences.
+
+            # Slicing ``episode_end_index`` is used to create a view of shape
+            # ([B, T, R + 1]) but each sequence (b, t) is filled with the values
+            # at the episode end, or at the sequence end if it does not cross
+            # the episode boundary. Such view is used to facilitate the helper
+            # function _unfold_dim1_adapting_episode_ends
+            episode_end_index = (torch.arange(B)[:, None, None],
+                                 torch.arange(T)[:, None],
+                                 torch.min(
+                                     steps_to_episode_end,
+                                     torch.full_like(steps_to_episode_end,
+                                                     R))[:, :, None].repeat(
+                                                         (1, 1, R + 1)))
+
+            def _unfold_dim1_adapting_episode_ends(x):
+                # Do the unfolding first.
+                y = _unfold_dim1(x)
+
+                # Fill the beyond episode end locations with the values from the
+                # aforementioned view.
+                y[beyond_episode_end] = y[episode_end_index][
+                    beyond_episode_end]
+                return y
```

Comment:
Can be simplified as:
```python
    capped_index = (torch.arange(B)[:, None, None],
                    torch.arange(T)[:, None],
                    torch.min(steps_to_episode_end.unsqueeze(-1), torch.arange(R+1)))
    def _unfold_dim1_adapting_episode_ends(x):
        return _unfold_dim1(y)[capped_index]
```
And it's possible to further combine `_unfold_dim1()` and `capped_index` as one single indexing operation `y[some_index]` where `some_index` takes into unfold and capped_index into account.

Reply:
Much nicer! Literally scratched my head trying to find a simpler way to achieve this. Thanks!